File: TopCoder/src/main/java/com/leetcode/ReverseNodesInKgroup.java
package com.leetcode;

//https://leetcode.com/problems/reverse-nodes-in-k-group/
public class ReverseNodesInKgroup {

    SwapNodesInPair.ListNode nextStart = null;
    public SwapNodesInPair.ListNode reverseKGroup(SwapNodesInPair.ListNode head, int k) {
        SwapNodesInPair.ListNode tail = head;

        for ( int i = 0 ; i < k - 1 ; i++) {
            tail = tail.next;
            if (tail == null) {
                return head;
            }
        }

        SwapNodesInPair.ListNode next = tail.next;
        tail.next = null;
        reverse(head);
        head.next = reverseKGroup(next, k);
        return tail;
    }

    private SwapNodesInPair.ListNode reverse(SwapNodesInPair.ListNode head) {
        SwapNodesInPair.ListNode prev = null;
        SwapNodesInPair.ListNode next = null;
        while (head != null) {
            next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }
        return prev;

    }



    public static void main(String[] args) {
        SwapNodesInPair.ListNode node0 = new SwapNodesInPair.ListNode(1);
        node0.next = new SwapNodesInPair.ListNode(2);
        node0.next.next = new SwapNodesInPair.ListNode(3);
        node0.next.next.next = new SwapNodesInPair.ListNode(4);
//
        node0.next.next.next.next = new SwapNodesInPair.ListNode(5);
//        node0.next.next.next.next.next = new SwapNodesInPair.ListNode(6);
//        node0.next.next.next.next.next.next = new SwapNodesInPair.ListNode(7);
//        node0.next.next.next.next.next.next.next = new SwapNodesInPair.ListNode(8);
//        node0.next.next.next.next.next.next.next.next = new SwapNodesInPair.ListNode(9);

        SwapNodesInPair.ListNode newHead = new ReverseNodesInKgroup().reverseKGroup(node0, 2);
        while(newHead != null) {
            System.out.println(newHead.val);
            newHead = newHead.next;
        }

    }
}

File: TopCoder/src/main/java/com/leetcode/ReverseNodesInKgroup.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:

    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        """
        prev_node = ListNode(-1)
        current = ListNode(-1)
        while head:
            new_head, next_start = self.reverse(head, k)
            if current.val == -1:
                current = new_head
            prev_node.next = new_head
            prev_node = head
            head = next_start

        if current.val == -1:
            return  None
        else:
            return current
        """
        dummy = jump = ListNode(0)
        dummy.next = l = r = head
        while 1:
            count = 0
            while r and count < k:
                r = r.next
                count += 1
            pre, cur = r, l
            if count == k:
                for _ in range(k):
                    cur.next, cur, pre = pre, cur.next, cur
                jump.next, jump, l = pre, l, r
            else:
                return dummy.next


    def reverse(self, head, k):
        if not head:
            return None, None

        if not head.next:
            return head, None

        if k == 1:
            return head, head.next

        curr_next = head
        prev_node = None
        count = 1
        while count <= k:
            if curr_next:
                curr_next = head.next
                head.next = prev_node
                prev_node = head
                head = curr_next
            else:
                return self.reverse(prev_node, count-1)

            count = count + 1
        return prev_node, curr_next

if __name__ == '__main__':
    node0 = ListNode(1)
    node0.next = ListNode(2)
    node0.next.next = ListNode(3)
    node0.next.next.next = ListNode(4)
    node0.next.next.next.next = ListNode(5)

    sol = Solution()
    ret = sol.reverseKGroup(node0, 2)

    while ret:
        print(ret.val)
        ret = ret.next
